Multiple inputs & results

TheLastGimbus opened the pull request:
Hi there!

I was looking for something like `@functools.lru_cache` from Python in Dart, and came across this library. But it's currently remebering only one result - kinda useless in my opinion :confused: 

### So I'm proposing a solution to write down results with a `HashMap`

Currently I added only `memo1`, but if you like it, I will extend it to all the others - probably nested `HashMap<A, HashMap<...>>` will be used for that

### I also added an example code to demonstrate how it makes things way faster :rocket: 

Closes #10

TheLastGimbus committed:
Init example

TheLastGimbus committed:
Add some heavy calculations example

TheLastGimbus committed:
First "multi input" function

TheLastGimbus commented:
Hey @Aetet @brianegan, is this repo alive or something? I can make a fork/re-make of this myself...

Aetet commented:
Hey.
Currently I'm not maintaining this repo. Some guys from Wrike company should.

@DisDis @Chudesnov @bunopus

sinegovsky-ivan commented on lib/memoize.dart:
```diff
@@ -20,19 +22,14 @@ Func0<R> memo0<R>(Func0<R> func) {
-  A prevArg;
-  R prevResult;
-  bool isInitial = true;
+  HashMap<A, R> argsToOutput = HashMap();
```
How about limit storage capacity? For example second argument (with default value) to memo1.

Otherwise, in some cases, we will waste a lot of memory.

TheLastGimbus replied:
Oh, you mean to limit maximum cache? Yeah, that would be useful

Tho i'm not sure how to do this - if you just `pop()` one item from a `HashMap` in dart, does it do it in some order or just random?

sinegovsky-ivan replied:
Better refer to the documentation https://api.dart.dev/stable/2.12.2/dart-collection/HashMap/HashMap.html :)

TheLastGimbus replied:
Okay, so generally - you like/want my change in general sense? If so, I can implement rest of the features - just wasn't sure if I even should

sinegovsky-ivan replied:
Yes, I like your changes:)

missvalentinep commented:
@TheLastGimbus a gentle ping :)

TheLastGimbus committed:
Merge branch 'master' into multiple_results

# Conflicts:
#	lib/memoize.dart

TheLastGimbus committed:
Write first function once again

TheLastGimbus committed:
Update example

TheLastGimbus committed:
A way to support multiple arguments

...I hope

TheLastGimbus committed:
Implement rest of functions

I hope ;_;

TheLastGimbus committed:
Make hashCode more compact

TheLastGimbus committed:
Shorter arg names to fit when auto-formatting

TheLastGimbus committed:
Fix tests... I hope :grimacing:

TheLastGimbus commented:
# HIiiiiii!

I know it took *some time*, but here it is! I updated branch with `master`, and implemented rest of methods! 

I also fixed (I hope that's the right word :sweat_smile:) test - they were designed that function re-executes every time *something* changes - please take a look at 84c2854

@sinegovsky-ivan I don't have any good idea for limiting the capacity tho :confused: It's a `HashMap`, so we can't just pop the first one (to make some space) because it's un-ordered... maybe we can store not only *the value* but also *some* `N` counter which by we would decide?

```dart
Func1<A, R> memo1<A, R>(Func1<A, R> func, {int? maxSize}) {
  final argsToOutput = HashMap<List, R>(
    equals: (a, b) => _listEquals(a, b),
    hashCode: (l) => _listHashCode(l),
  );
  // External list to keep track of cache order
  final hashes = <int>[];

  return ((A aA) {
    final cached = argsToOutput[[aA]];
    if (cached != null) {
      return cached;
    } else {
      final res = func(aA);
      argsToOutput[[aA]] = res;
      // Skip the whole thing if user doesn't use it
      if (maxSize != null) {
        hashes.add(_listHashCode([aA]));
        if (argsToOutput.length > maxSize) {
          // Pop the first (oldest) one
          // I'm not sure how efficient this is :/
          argsToOutput.removeWhere((k, _) => _listHashCode(k) == hashes[0]);
          hashes.removeAt(0);
        }
      }
      return res;
    }
  });
}
```
Something like this?

Edit: I actually wrote simple test for the above:
```dart
test('should remove old cache when maxSize set', () {
  var count = 0;
  var func = memo1((int a) => ++count, maxSize: 2);
  expect(func(1), 1);
  expect(func(1), 1);
  expect(func(2), 2);
  expect(func(2), 2);
  expect(func(1), 1);
  expect(func(3), 3);
  // At this point, first cache element should be removed
  expect(func(1), 4);
});
```
and it passes! 

...oh, and also - I didn't know what to do with `imemoN` functions - they compare by `identical`, and I'm not sure will this work in `HashMap` - but I can try...

missvalentinep commented:
@TheLastGimbus hey! I come to you with many thoughts about this PR:
1. Creating a hash map for every memo is not very efficient (since we'll just increase memory usage for people who're already using these functions) and adding an "if" condition ("create hashmap only if maxSize was provided") would make the code hard to read. So I propose creating a separate set of functions (just like imemo), called **memom1, memom2...** (as in "memoMany") that will have this functionality you want to add.
2. As for `_listEquals` and `_listHashCode`, I'd propose you use **quiver** package and `listEqual` and `hashObjects` methods from there
3. For storing cache I believe you should use [LinkedHashMap](https://api.dart.dev/stable/2.13.4/dart-collection/LinkedHashMap-class.html), as it remembers the order in which the keys were inserted. So it will allow you to do something like this:
```
if (argsToOutput.length == maxSize) {
  argsToOutput.remove(argsToOutput.keys.first);
}
```
4. Since we're creating a separate set of functions, the `maxSize` param should be mandatory.
5. [Optional] I'd personally rename `argsToOutput` to `cachedResults`/`cache` and `maxSize` to `cacheLimit`

om-ha commented:
Great project and PR!

Thanks for your review @missvalentinep.

Thanks @TheLastGimbus for your PR, would love to see this merged.
